### ISSUE TITLE
[Feature Request] Efficient Computation of Jacobian products JJ^T

### ISSUE THREAD

**Issue by activatedgeek** [wilson-labs/cola#17]
# 🚀 Feature Request

I recently came across a neat trick to compute Jacobian matrix product of the form JJ^T using VJPs and JVPs. I think this would make for a good addition as a utility in the library, very obvious but probably useful function to have. Let me know what you think!

## Motivation

**Is your feature request related to a problem? Please describe.**

The matrix product JJ^T comes up in probit classification, where the posterior predictive integral of the usual softmax classification is approximated via a linearization of the pre-softmax function. It also comes up is Function-Space VI. I'm sure there are other places it comes up.

The key idea: For a matrix J of size `M x N` (M outputs, N inputs/parameters), such that `M` is much smaller than `N`, it is useful construct the symmetric product JJ^T column-by-column. First constructing a VJP `J^T v` and then JVP `J(J^T v)`. We would never have a larger `M x N` matrix, but only a smaller `M x M` matrix when the vectors `v` are one-hot vectors of size `M`.

PyTorch `vmap` also allows chunking operations which might perhaps be helpful to compute this block-by-block to avoid OOM constraints. Unfortunately, I could not find a chunking operation for `jax.vmap` but perhaps that can be manually implemented.

## Pitch

**Describe the solution you'd like**

I'd like a utility function be exposed from cola, something to the effect of:
```
import cola

model = ## pytorch model
f_model = ## a torch.func version of the model, that handles single input and single output.

cola.jjt(f_model)
```
which computes the JJ^T. There are a few design considerations here, as to how one handles the batches. Batches should perhaps not be within the cola scope, but rather the user's responsibility.

**Describe alternatives you've considered**

Here's a reference toy implementation in PyTorch to give you a sense of what would need to be implemented.

```python
import torch
import torch.nn as nn
import torch.func as tf

model = nn.Sequential(nn.Linear(2, 10), nn.ReLU(), nn.Linear(10, 5))
f_model = lambda p, *args, **kwargs: tf.functional_call(model, p, *args, **kwargs)

inputs = torch.randn(100, 2)
params = dict(model.named_parameters())
outputs_size = torch.Size([5])  ## Only handles 1-D outputs

## Naive computation using tf.jacrev/tf.jacfwd that constructs full Jacobian. vmap over batch size.
j_fn_naive = tf.vmap(
    lambda J_p: torch.cat([v.view(*outputs_size, -1) for _, v in J_p.items()], dim=-1)
)
jjt_fn_naive = tf.vmap(lambda M: M @ M.T)
jjt_naive = jjt_fn_naive(j_fn_naive(tf.jacrev(f_model)(params, inputs)))

## Smarter computation. Outer vmap over batch size, inner vmap over one-hot co-tangent vectors.
jjt_fn = tf.vmap(
    lambda inp: tf.vmap(
        lambda v: tf.jvp(
            lambda p: f_model(p, inp.unsqueeze(0)).squeeze(0),
            (params,),
            tf.vjp(lambda p: f_model(p, inp.unsqueeze(0)).squeeze(0), params)[1](v),
        )[1]
    )(torch.eye(*outputs_size))
)
jjt = jjt_fn(inputs)

print(jjt.shape, jjt_naive.shape)
assert torch.allclose(jjt, jjt_naive)
```

**Are you willing to open a pull request?** (We LOVE contributions!!!)

Sure! Once we have some design decisions figured out.

**Comment by mfinzi** [comment]
Great idea. And thanks for using the template!
I think that should be possible in principle now with something like

```
p = flatten(params)
J = cola.ops.Jacobian(f_model, p)
JJT = J@J.T
```

though we will need to make sure we have the right version of pytorch, since earlier versions the torch.func.vmap was not working properly
(it's possible the [torch_fns.vmap](https://github.com/wilson-labs/cola/blob/main/cola/torch_fns.py) needs to be amended from when we wrote it with a previous version)

Let's compare the pure runtime of `J@J.T` with the matmat you listed above, they may be quite similar.

**Comment by activatedgeek** [comment]
Ah got it. 

So is `J @ J.T` already implemented via VJPs/JVPs internally? I only searched to docs for some reference to the JJ^T product but wasn't sure about the implementation.

My PyTorch version was 2.0.1, but I think the public-API of `torch.func` is relatively stable now?

It looks like you are using some internal API `torch._vmap_internals` and perhaps need to just use `torch.func.vmap` now.

**Comment by mfinzi** [comment]
See [07_second_order_optimization.html](https://cola.readthedocs.io/en/latest/notebooks/07_second_order_optimization.html) for example of using J@J.T in pytorch. I think the jjt implementation above may be faster?, but we should test it

**Comment by mfinzi** [comment]
Circling back to this, perhaps we should try to get a version that does chunking as the current `Jacobian` is just using the direct vmap
